lib/IDE/CompletionInstance.cpp

```diff
@@ -162,10 +165,120 @@ static DeclContext *getEquivalentDeclContextFromSourceFile(DeclContext *DC,
   return newDC;
 }
 
+/// For each dependency files in CI, run \p callback until the callback returns
+/// \c true. Returns \c true if any callback call returns \c true, \c false
+/// otherwise.
+bool forEachDependenciesUntilTrue(
+    CompilerInstance &CI, ModuleDecl *CurrentModule, unsigned excludeBufferID,
+    llvm::function_ref<bool(StringRef)> callback) {
+  // Check files in the current module.
+  for (FileUnit *file : CurrentModule->getFiles()) {
+    StringRef filename;
+    if (auto SF = dyn_cast<SourceFile>(file)) {
+      if (SF->getBufferID() == excludeBufferID)
+        continue;
+      filename = SF->getFilename();
+    } else if (auto LF = dyn_cast<LoadedFile>(file))
+      filename = LF->getFilename();
+    else
+      continue;
+
+    // Ignore synthesized files.
+    if (filename.empty() || filename.front() == '<')
+      continue;
+
+    if (callback(filename))
+      return true;
+  }
+
+  // Check other non-system depenencies (e.g. modules, headers).
+  for (auto &dep : CI.getDependencyTracker()->getDependencies()) {
+    if (callback(dep))
+      return true;
+  }
+
+  return false;
+}
+
+/// Collect hash codes of the dependencies into \c Map.
+/// We need to do this only if the file system is in-mememory.
+void cacheDependencyHashIfNeeded(CompilerInstance &CI,
+                                 ModuleDecl *CurrentModule,
+                                 unsigned excludeBufferID,
+                                 llvm::StringMap<llvm::hash_code> &Map) {
+  auto &FS = CI.getFileSystem();
+  forEachDependenciesUntilTrue(
+      CI, CurrentModule, excludeBufferID, [&](StringRef filename) {
+        if (Map.count(filename))
+          return false;
+
+        auto stat = FS.status(filename);
+        if (!stat)
+          return false;
+
+        // Stop processing if the modification time is not zero. Probably this
+        // is not a virtual file system.
+        if (stat->getLastModificationTime() != llvm::sys::TimePoint<>())
+          return true;
```

Comment:
I think I'm not understanding something: why do we want to stop as soon as we see the first "normal file"? Why can't we have virtual mixed in with regular?

Reply:
I just want to save the `stat` time if the filesystem is not in-memory.
AFAIU, in reality, we don't mix in-memory file with normal file.

Reply:
The filesystem you get from our builtin FileSystemProvider in sourcekitd is an overlay, so if a call is not handled by the InMemoryFilesystem, it will fall through to the RealFilesystem.

Reply:
Ah, thank you for notifying! Updated to just ignore the non-zero modification time here.